src/algorithms/particle/CaloRemnantCombiner.cc

```diff
@@ -0,0 +1,154 @@
+
+#include "algorithms/particle/CaloRemnantCombinerConfig.h"
+
+#include "CaloRemnantCombiner.h"
+
+namespace eicrecon {
+
+void CaloRemnantCombiner::process(const CaloRemnantCombiner::Input& input,
+                                  const CaloRemnantCombiner::Output& output) const {
+
+  const auto [calo_clusters]    = input;
+  auto [out_neutral_candidates] = output;
+
+  std::vector<bool> visits_ecal(calo_clusters[0]->size(), false);
+  std::vector<bool> visits_hcal(calo_clusters[1]->size(), false);
+
+  std::vector<bool> visits_ecal_true(calo_clusters[0]->size(), true);
+  std::vector<bool> visits_hcal_true(calo_clusters[1]->size(), true);
+
+  while (visits_ecal != visits_ecal_true) {
+
+    edm4eic::MutableReconstructedParticle neutral_candidate_eh;
+
+    // Step 1: Find the seed Ecal cluster with highest energy
+    std::size_t seed_ecal_index = findSeedCluster_index(*calo_clusters[0], visits_ecal);
+
+    if (seed_ecal_index == -1) {
+      info("No Seed Ecal cluster found for remnant combination.");
+    }
+
+    if (seed_ecal_index != -1) {
+      // Get the cluster indices for merging
+      std::set<std::size_t> ecalcluster_indices = getcluster_indices_for_merging(
+          *calo_clusters[0], visits_ecal, seed_ecal_index, m_cfg.delta_r_add_em, *calo_clusters[0]);
+
+      for (const auto& idx : ecalcluster_indices) {
+        neutral_candidate_eh.addToClusters((*calo_clusters[0])[idx]);
+      }
+
+      std::set<std::size_t> e_hcalcluster_indices = getcluster_indices_for_merging(
+          *calo_clusters[1], visits_hcal, seed_ecal_index, m_cfg.delta_r_add_h, *calo_clusters[0]);
+
+      for (const auto& idx : e_hcalcluster_indices) {
+        neutral_candidate_eh.addToClusters((*calo_clusters[1])[idx]);
+      }
+
+      out_neutral_candidates->push_back(neutral_candidate_eh);
+
+    } // end of if (seed_ecal_index != -1)
+
+  } // end of while (visits_ecal != visits_ecal_true)
+
+  while (visits_hcal != visits_hcal_true) {
+
+    edm4eic::MutableReconstructedParticle neutral_candidate_h;
+    std::size_t seed_rem_hcal_index = findSeedCluster_index(*calo_clusters[1], visits_hcal);
+
+    if (seed_rem_hcal_index == -1) {
+      info("No Seed Hcal cluster found for remnant combination.");
+    }
+
+    if (seed_rem_hcal_index != -1) {
+
+      std::set<std::size_t> rem_hcalcluster_indices;
+
+      rem_hcalcluster_indices =
+          getcluster_indices_for_merging(*calo_clusters[1], visits_hcal, seed_rem_hcal_index,
+                                         m_cfg.delta_r_add_h, *calo_clusters[1]);
+
+      for (const auto& idx : rem_hcalcluster_indices) {
+        neutral_candidate_h.addToClusters((*calo_clusters[1])[idx]);
+      }
+      out_neutral_candidates->push_back(neutral_candidate_h);
+
+    } // end of if (seed_rem_hcal_index != -1)
+
+  } // end of while (visits_hcal != visits_hcal_true)
+
+} // end of process
+
+std::size_t CaloRemnantCombiner::findSeedCluster_index(const edm4eic::ClusterCollection& clusters,
```

Comment:
Another good place for a docstring! Here you'll definitely want to note the criteria for a seed cluster and what it's for.
```suggestion
// ----------------------------------------------------------------------------
//! Find seed cluster
// ----------------------------------------------------------------------------
/*! Identifies a seed (highest energy) cluster in a collection 
 *  which sets the center of the cone in which clusters are
 *  combined.
 */
std::size_t CaloRemnantCombiner::find_seed_cluster_index(const edm4eic::ClusterCollection& clusters,
```